[UI-side compositing] Have RemoteScrollbarsController keep track of scrollbar visibility for hit testing

nmoucht opened the pull request:
#### a9d8f8b89517139a774a25647ffcbf0e44fc6b8c
<pre>
[UI-side compositing] Have RemoteScrollbarsController keep track of scrollbar visibility for hit testing
<a href="https://bugs.webkit.org/show_bug.cgi?id=255303">https://bugs.webkit.org/show_bug.cgi?id=255303</a>
rdar://107859102

Reviewed by Simon Fraser.

Overlay scrollbars should only participate in hit testing when they are visible. When we do a knob alpha
animation in the UI process, send a visibility change event to the web process, notifying
RemoteScrollbarsController so it can properly report in shouldScrollbarParticipateInHitTesting
whether the scrollbar should participate in hit testing or not.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::scrollbarVisibilityDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollerImpDelegateMac scrollerImp:animateKnobAlphaTo:duration:]):
(WebCore::ScrollerMac::visibilityChanged):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
(WebCore::ScrollerPairMac::scrollingNode):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::setScrollbarVisibilityState):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm: Renamed from Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.cpp.
(WebKit::RemoteScrollbarsController::RemoteScrollbarsController):
(WebKit::RemoteScrollbarsController::mouseEnteredContentArea):
(WebKit::RemoteScrollbarsController::mouseExitedContentArea):
(WebKit::RemoteScrollbarsController::mouseMovedInContentArea):
(WebKit::RemoteScrollbarsController::mouseEnteredScrollbar const):
(WebKit::RemoteScrollbarsController::mouseExitedScrollbar const):
(WebKit::RemoteScrollbarsController::shouldScrollbarParticipateInHitTesting):
(WebKit::RemoteScrollbarsController::setScrollbarVisibilityState):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::scrollingTreeNodeScrollbarVisibilityDidChange):

Canonical link: <a href="https://commits.webkit.org/263004@main">https://commits.webkit.org/263004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3075b626068395cbfd2c921e6d4a7dd3c7f701ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3264 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3322 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3433 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4674 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3592 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3234 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3376 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3343 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2828 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3298 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3587 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4494 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/3289 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1088 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2900 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2817 "1 flakes 142 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2870 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2953 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4227 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3333 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2659 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2901 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2895 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/386 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3174 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
```diff
@@ -69,5 +70,23 @@ void RemoteScrollbarsController::mouseExitedScrollbar(WebCore::Scrollbar* scroll
+bool RemoteScrollbarsController::shouldScrollbarParticipateInHitTesting(WebCore::Scrollbar* scrollbar)
+{
+    // Non-overlay scrollbars should always participate in hit testing.
+    if (ScrollerStyle::recommendedScrollerStyle() != NSScrollerStyleOverlay)
```
We should get the scrollbar style from the the `scrollbar` itself, not from `recommendedScrollerStyle()`, because this might be a custom scrollbar.

nmoucht replied:
This is checked in Scrollbar::shouldParticipateInHitTesting.

smfr replied:
So can we just `ASSERT(scrollbar->isOverlayScrollbar())` here then? That’s less confusing.

nmoucht replied:
Right, fixed.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
```diff
@@ -46,8 +46,13 @@ class RemoteScrollbarsController final : public WebCore::ScrollbarsController {
+    void setScrollbarVisibilityState(ScrollbarOrientation, bool) final;
 
 private:
+    bool m_horizontalScrollbarIsVisible { false };
```
We don't initialize these to the correct values for non-overlay scrollbars, so they aren't always correct, and there's no code to handle toggles between overlay and non-overlay scrollbars when the user changes the system scrollbar style.

smfr replied:
I suggest you either name these to make it explicit that they only apply to overlay scrollbars, or make sure the state is always correct.

nmoucht replied:
Right, renamed to specify this is only for overlay scrollbars.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
```diff
@@ -96,6 +96,7 @@ class RemoteScrollingCoordinator final : public WebCore::AsyncScrollingCoordinat
     void receivedWheelEventWithPhases(WebCore::PlatformWheelEventPhase phase, WebCore::PlatformWheelEventPhase momentumPhase);
     void startDeferringScrollingTestCompletionForNode(WebCore::ScrollingNodeID, WebCore::WheelEventTestMonitor::DeferReason);
     void stopDeferringScrollingTestCompletionForNode(WebCore::ScrollingNodeID, WebCore::WheelEventTestMonitor::DeferReason);
+    void scrollingTreeNodeScrollbarVisibilityDidChange(uint64_t, WebCore::ScrollbarOrientation, bool);
```
This should use `WebCore::ScrollingNodeID`, not `uint64_t`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
```diff
@@ -185,6 +186,11 @@
     return WheelEventHandlingResult::handled();
 }
 
+void RemoteScrollingCoordinator::scrollingTreeNodeScrollbarVisibilityDidChange(uint64_t nodeID, WebCore::ScrollbarOrientation orientation, bool isVisible)
```
I don't think you need this override at all?

nmoucht replied:
Since this needs to be a RemoteScrollingCoordinator function for the IPC, I dont think there’s a way to get around having this function.

smfr replied:
OK. The first argument should be a ScrollingNodeID.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
```diff
@@ -30,6 +30,7 @@
 
 #import "ArgumentCoders.h"
 #import "GraphicsLayerCARemote.h"
+#import "Logging.h"
```
Unused?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.h:
```diff
@@ -69,8 +69,10 @@ class ScrollerMac {
+    void visibilityChanged(bool);
 
 private:
+    bool m_isVisible;
```
Should have an initializer.

nmoucht replied:
Fixed.

nmoucht committed:
[UI-side compositing] Have RemoteScrollbarsController keep track of scrollbar visibility for hit testing
https://bugs.webkit.org/show_bug.cgi?id=255303
rdar://107859102

Reviewed by Simon Fraser.

Overlay scrollbars should only participate in hit testing when they are visible. When we do a knob alpha
animation in the UI process, send a visibility change event to the web process, notifying
RemoteScrollbarsController so it can properly report in shouldScrollbarParticipateInHitTesting
whether the scrollbar should participate in hit testing or not.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::scrollbarVisibilityDidChange):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollerImpDelegateMac scrollerImp:animateKnobAlphaTo:duration:]):
(WebCore::ScrollerMac::visibilityChanged):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
(WebCore::ScrollerPairMac::scrollingNode):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollbarsController.h:
(WebCore::ScrollbarsController::setScrollbarVisibilityState):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeScrollbarVisibilityDidChange):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm: Renamed from Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.cpp.
(WebKit::RemoteScrollbarsController::RemoteScrollbarsController):
(WebKit::RemoteScrollbarsController::mouseEnteredContentArea):
(WebKit::RemoteScrollbarsController::mouseExitedContentArea):
(WebKit::RemoteScrollbarsController::mouseMovedInContentArea):
(WebKit::RemoteScrollbarsController::mouseEnteredScrollbar const):
(WebKit::RemoteScrollbarsController::mouseExitedScrollbar const):
(WebKit::RemoteScrollbarsController::shouldScrollbarParticipateInHitTesting):
(WebKit::RemoteScrollbarsController::setScrollbarVisibilityState):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::scrollingTreeNodeScrollbarVisibilityDidChange):

Canonical link: https://commits.webkit.org/263004@main
